### ISSUE TITLE
[v2] isSubmitting resets to false too early.

### ISSUE THREAD

**Issue by rafaelderolez** [jaredpalmer/formik#1730]
### Current Behavior
In 2.0.1-rc12, when submitting the form, isSubmitting is set to `true` but is instantly reset to `false`, making it hard to disable input fields and buttons.

### Expected behavior
I would expect `isSubmitting` to be true until it's set to false in the onSubmit function.

### Reproducible example
[Codesandbox example.](https://codesandbox.io/s/formik-codesandbox-issubmitting-rguft)

### Your environment
| Software         | Version(s) |
| ---------------- | ---------- |
| Formik           | 2.0.1-rc12
| React            | 16.8.6
| Browser          | Google Chrome  Version 76.0.3809.87 (64-bit)
| npm/Yarn         | 1.17.3
| Operating System | macOS 10.14.5

**Comment by schrapel** [comment]
I just ran into this issue as well.

**Comment by jaredpalmer** [comment]
Confirmed

**Comment by karlandin** [comment]
I ran into the issue as well and noticed that it works as I would expect if my onSubmit function returns a promise. isSubmitting is set to false first when the promise resolves.

**Comment by shulcsm** [comment]
What is the expected here?
Toggling state on promise resolution makes sense but signature of submitHandler is `(values: Values, formikBag: FormikBag) => void`.
Perhaps it should also accept `(values: Values, formikBag: FormikBag) => Promise<T>` and do what it does now?

**Comment by Ericnr** [comment]
Ran into this problem when implementing modal forms too in #1741.

**Comment by jaredpalmer** [comment]
Thanks for the write up. Will investigate

**Comment by mhienle** [comment]
Observing the same issue in v1.5.8

**Comment by mhienle** [comment]
I had a similar issue in v1.5.8 which I was able to resolve by setting `enableReinitialize` to its default value `false`.

**Comment by alexborisov** [comment]
any updates on this issue? I'm having the same problem and it's causing issues disabling the submit button for me - breaking functionality. Setting `enableReinitialize` to `false` as suggested by @mhienle did not help :(

**Comment by danecando** [comment]
@alexborisov did you try `return await callApi()` 

As someone else suggested in my case my operation was async and I wasn't returning the promise. Seems like that would even be the intended behavior. If the onSubmit method doesn't return a promise it would assume it's done with the work.

**Comment by alexborisov** [comment]
@danecando yup returning async works for me. As does returning a promise. Thanks!

**Comment by jaredpalmer** [comment]
Do we need to update the docs around this?

**Comment by jaredpalmer** [comment]
Perhaps in the tutorial as well?

**Comment by danecando** [comment]
@jaredpalmer I believe it would be worth mentioning. I think in a previous version it worked fine with me handling the submit state manually in my async callback but it didn't work that way for me on v2. Not sure if this is new behavior or my memory sucks but returning the promise seems like a simpler api to me anyways.

Let me know I'll be happy to submit a pr sometime this week mentioning this behavior in the docs. PS: v2 is **the :bomb:** great work jared and contributors.

**Comment by RodyGL** [comment]
Here is an example of the issue. Just enter an email and click as many times as you can within 1 sec. See how **submitCount** increases.

https://codesandbox.io/s/formik-example-67mpu?fontsize=14

**Comment by jaburcodes** [comment]
Also facing this issue

**Comment by michielmetcake** [comment]
So returning async seems to work.. 

But how to handle dispatching redux actions on submit?

**Comment by omarshammas** [comment]
Returning the promise solved my issue in case anyone else faces this issue.

**Comment by rofazayn** [comment]
I am still getting the same problem, any solutions please?

**Comment by willxy** [comment]
Related to #1957 and #1987 sounds like

**Comment by sibelius** [comment]
You should follow this approach https://github.com/relayjs/relay-examples/pull/118

you need both, ref and state to avoid double clicks/submits

**Comment by kevinvugts** [comment]
> Confirmed

Hi,

I just experienced this on "formik": "^2.0.4" as well. What is the fix for this?

**Comment by jaredpalmer** [comment]
What's the suggested solution here?

**Comment by sibelius** [comment]
also this https://twitter.com/davidkpiano/status/1194616763000008705?s=21

we can return `isSubmitting` to user (to use as disabled on buttons), and use a `ref` to avoid double `handleSubmit` calls (this ref could be only for internal use)

**Comment by j0b0sapi3n** [comment]
@jaredpalmer #1987 is the solution many of us are hoping would be accepted which would allow many of us to not have to change the way we currently handle `isSubmitting` in `v1.5.8`. However, we're unsure if this is something you actually support. Otherwise, many of us have reverted to `v1.5.8` for now. Thanks for taking a look at this!

**Comment by kush3107** [comment]
So now that the isSubmitting state of Formik is controlled by the promise return in OnSubmit handler don't you think that exposing the `setSubmitting` method is stale.

**Comment by gustavopch** [comment]
> @jaredpalmer So now that the isSubmitting state of Formik is controlled by the promise return in OnSubmit handler don't you think that exposing the `setSubmitting` method is stale.

It's not. As the release note of v2.0.5 says:
> HOWEVER, if your onSubmit function is synchronous (e.g. v1), then you need to still call setSubmitting(false) on your own

**Comment by charleskoehl** [comment]
> I ran into the issue as well and noticed that it works as I would expect if my onSubmit function returns a promise. isSubmitting is set to false first when the promise resolves.

I've tried a few ways to get my onSubmit function to return a promise but isSubmitting is still set to false immediately. How are you doing it?

**Comment by sgnl** [comment]
updating the documentation to mention that "Returning the promise" will give `isSubmitting` a better, not flakey value of the form's submittion state - would be awesome

> So returning async seems to work..
> 
> But how to handle dispatching redux actions on submit?

Have you tried something like:

```
const onSubmit={() => {
    return new Promise((resolve, reject) => {
      displayYourActions({type: '', value: [], resolve, reject, ...anyOtherData?})
      // Make sure your action calls resolve or reject!
    })
  }}
```

**Comment by namelesswc** [comment]
I use redux-thunk and this works for me. Make sure your action returns a promise aswell!

```
  const handleSubmit = (values, {setSubmitting}) => {
    setSubmitting(true);
    return dispatch(myAction(values))
      .then(() => {
        setSubmitting(false);
        return Promise.resolve();
      })
  }
```

**Comment by kaluabentes-hotmart** [comment]
> > @jaredpalmer So now that the isSubmitting state of Formik is controlled by the promise return in OnSubmit handler don't you think that exposing the `setSubmitting` method is stale.
> 
> It's not. As the release note of v2.0.5 says:
> 
> > HOWEVER, if your onSubmit function is synchronous (e.g. v1), then you need to still call setSubmitting(false) on your own

But still not works

**Comment by Gowdhampr** [comment]
If onSubmit is async, then Formik will automatically set isSubmitting to false on your behalf once it has resolved. This means you do NOT need to call formikBag.setSubmitting(false) manually. However, if your onSubmit function is synchronous, then you need to call setSubmitting(false) on your own.

In my case, onSubmit was async but was using a react query mutate function for api call, so i just removed it and it solved the problem

Refer [Doc](https://formik.org/docs/api/formik#onsubmit-values-values-formikbag-formikbag--void--promiseany:~:text=IMPORTANT%3A%20If%20onSubmit,on%20your%20own.)

**Comment by TaylorFoster-Informulate** [comment]
**For anyone coming across this in 2024 or later**

I beat my head against the wall for days with this issue. I have a form, the submit handler had multiple promises in it, and isSubmitting was setting back to `false` after the first promise. I could't figure out why. 

```
const onSubmit = async () => {
  try { 
    await one();
    // isSubmitting gets set back to true here, not after two()
    await two();
  } catch(e) {
    console.error(e);
  }
}
```

**The Problem**
The issue here is that `await one();` was making a call to my API which was updating data that was **directly related** to the form. My `initialValues` in the form contained references to the same data I was updating. Because of this formik says: "Ok ill submit, my boolean is now true, oh look my initialState just changed, I need to reinitialize myself which means I can no longer be in a submitting state, im starting fresh". 

**To Solve**
You need to remove `enableReinitialize` from your `<Formik>` prop. This will tell Formik to stop "reseting" the form after its dependencies (in this case initial values) change. Removing this will fix your problem. 

**Caution**
Disabling `enableReinitialize` can cause some odd behavior in your form and sometimes you do in fact want the form to reinit when a piece of outside state changes. There is plenty of use cases for that. But, you will run into [this problem here](https://github.com/jaredpalmer/formik/issues/811). 

I love formik, I even use it sometimes just for state management (not even as a form, kind of like a context wrapper). But this issue is hella anoying and there is no "good" way around it. 

Good luck folks!
